rust/src/treefile.rs

```diff
@@ -38,15 +38,35 @@ pub struct Treefile {
     pub rojig_spec: Option<Box<CStr>>,
 }
 
+enum InputFormat {
+    YAML,
+    JSON,
+}
+
 /// Parse a YAML treefile definition using architecture `arch`.
-fn treefile_parse_yaml<R: io::Read>(input: R, arch: Option<&str>) -> io::Result<TreeComposeConfig> {
-    let mut treefile: TreeComposeConfig = match serde_yaml::from_reader(input) {
-        Ok(t) => t,
-        Err(e) => {
-            return Err(io::Error::new(
-                io::ErrorKind::InvalidInput,
-                format!("serde: {}", e),
-            ))
+fn treefile_parse_stream(
+    fmt: InputFormat,
+    input: &mut io::Read,
```

Comment:
Minor: why the switch away from a generic here? Seems fine either way, just want to understand the motivation.

Reply:
There were like 2-3 rewrites/false starts here before this patch got split out as a prep for the bigger PR. I don't think there was any deliberate reason, and it does make sense to me to avoid vfuncs here.

Reply:
(will do a followup)